src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncDisposableTests.cs

```diff
@@ -0,0 +1,1379 @@
+            comp.VerifyDiagnostics(
+                // (6,15): error CS4033: The 'await' operator can only be used within an async method. Consider marking this method with the 'async' modifier and changing its return type to 'Task'.
+                //         using await (var x = new C())
+                Diagnostic(ErrorCode.ERR_BadAwaitWithoutVoidAsyncMethod, "await").WithLocation(6, 15)
+                );
+        }
+
+        [Fact]
+        public void TestInNonAsyncMethodReturningInt()
+        {
+            string source = @"
+class C : System.IAsyncDisposable
+{
+    int M()
+    {
+        using await (var x = new C())
+        {
+            return 1;
+        }
+    }
+    public System.Threading.Tasks.Task DisposeAsync()
+    {
+        throw null;
+    }
+}
+";
+            var comp = CreateCompilationWithMscorlib46(source + s_interfaces);
+            comp.VerifyDiagnostics(
+                // (6,15): error CS4032: The 'await' operator can only be used within an async method. Consider marking this method with the 'async' modifier and changing its return type to 'Task<int>'.
+                //         using await (var x = new C())
+                Diagnostic(ErrorCode.ERR_BadAwaitWithoutAsyncMethod, "await").WithArguments("int").WithLocation(6, 15)
+                );
+        }
+
+        [Fact]
+        public void TestInNonAsyncAnonymousMethod()
+        {
+            string source = @"
+class C : System.IAsyncDisposable
+{
+    void M()
+    {
+        System.Action x = () =>
+        {
+            using await (var y = new C())
+            {
+                return;
+            }
+        };
+    }
+    public System.Threading.Tasks.Task DisposeAsync()
+    {
+        throw null;
+    }
+}
+";
+            var comp = CreateCompilationWithMscorlib46(source + s_interfaces);
+            comp.VerifyDiagnostics(
+                // (8,19): error CS4034: The 'await' operator can only be used within an async lambda expression. Consider marking this lambda expression with the 'async' modifier.
+                //             using await (var y = new C())
+                Diagnostic(ErrorCode.ERR_BadAwaitWithoutAsyncLambda, "await").WithArguments("lambda expression").WithLocation(8, 19)
+                );
+        }
+
+        [Fact]
+        public void TestInAsyncAnonymousMethod()
+        {
+            string source = @"
+using System.Threading.Tasks;
+class C : System.IAsyncDisposable
+{
+    C()
+    {
+        System.Console.Write(""C "");
+    }
+    public static async Task Main()
+    {
+        System.Func<Task> x = async () =>
+        {
+            using await (var y = new C())
+            {
+                System.Console.Write(""body "");
+            }
+            System.Console.Write(""end"");
+        };
+        await x();
+    }
+    public Task DisposeAsync()
+    {
+        System.Console.Write(""DisposeAsync "");
+        return Task.Delay(10);
```

Comment:
>return Task.Delay(10); [](start = 8, length = 22)

How do we know that this task is awaited after this method is called?

Reply:
Here and in all other similar tests we should await for something inside DisposeAsync and then print something after that. This will confirm that we aren't just calling the method, but also awaiting the task that we are getting from it.

---
In reply to: [162766376](https://github.com/dotnet/roslyn/pull/23961#discussion_r162766376) [](ancestors = 162766376)

Reply:
Thanks for the clarification.